### PR TITLE
s3 backend management

### PR DESCRIPTION
**Pull request by Benjile**
* Generate terraform backend
* use sts to get temporary aksk from vault

### CODE REVIEW

File: tf [hunk 1]
```diff
@@ -175,16 +189,36 @@ function _tf_init () {
...
 
+function _tf_backend_config () {
+	if ${TF_USE_VAULT_AWS_STS}; then
+    token=$(vault read "${AWS_STS_PATH}")
+    log_debug "${token}"
+    TF_BACKEND_ACCESS_KEY=$(echo "$token" | jq -r '.data.access_key')
+    TF_BACKEND_SECRET_KEY=$(echo "$token" | jq -r '.data.secret_key')
+    TF_BACKEND_SESSION_TOKEN=$(echo "$token" | jq -r '.data.security_token')
+    # set backend config options
+    BACKEND_CONFIG="-backend-config=access_key=${TF_BACKEND_ACCESS_KEY} -backend-config=secret_key=${TF_BACKEND_SECRET_KEY} -backend-config=session_token=${TF_BACKEND_SESSION_TOKEN}"
```
Comment (eonpatapon) [inline tf:218]:
```suggestion
    echo "-backend-config=access_key=${TF_BACKEND_ACCESS_KEY} -backend-config=secret_key=${TF_BACKEND_SECRET_KEY} -backend-config=session_token=${TF_BACKEND_SESSION_TOKEN}"
```

File: tf [hunk 2]
```diff
@@ -175,16 +189,36 @@ function _tf_init () {
...
 
   # environment replacement in every *tf* files
   sed -i "s/#ENVIRONMENT#/${ENVIRONMENT}/g" "${TF_CONFIG_DIR}"/*.tf*
 
+  # get backend config
+	BACKEND_CONFIG=$(_tf_backend_config)
+
   # terraform init
-  _tf_generic init -upgrade=true
+  _tf_generic init -upgrade=true ${BACKEND_CONFIG}
```
Comment (eonpatapon) [inline tf:207]:
```suggestion
  _tf_generic init -upgrade=true $(_tf_backend_config)
```

File: tf [hunk 3]
```diff
@@ -175,16 +189,36 @@ function _tf_init () {
...
+	fi
+
   # add any tf and tfvars files present here to override the downloaded configuration
   cp ./*.{tf,tfvars,tfvars.json} "${TF_CONFIG_DIR}" &>/dev/null || true
 
   # environment replacement in every *tf* files
   sed -i "s/#ENVIRONMENT#/${ENVIRONMENT}/g" "${TF_CONFIG_DIR}"/*.tf*
 
+  # get backend config
+	BACKEND_CONFIG=$(_tf_backend_config)
```
Comment (eonpatapon) [inline tf:204]:
```suggestion
```

File: tf [hunk 4]
```diff
@@ -175,16 +189,36 @@ function _tf_init () {
   # fetch the lib repository
   _tf_fetch
 
+  # generate backend file
+	if ${TF_GENERATE_BACKEND}; then
+		_tf_generate_backend
+	fi
+
```
Comment (eonpatapon) [inline tf:196]: Indent is off

File: tf [hunk 5]
```diff
@@ -191,6 +237,19 @@ function _is_local () {
...
 }
 
+function _tf_generate_backend () {
+    cat <<EOF > "${TF_CONFIG_DIR}/backend.tf"
+# generated by tf
+terraform {
+  backend "s3" {
+    bucket = "${TF_BACKEND_BUCKET_NAME}"
+    key    = "${ENVIRONMENT}/${CONFIGURATION}/terraform.tfstate"
+    region = "${TF_BACKEND_BUCKET_REGION_NAME}"
```
Comment (yesteph) [inline tf:247]: Set dynamodb_table is not empty

File: tf [hunk 6]
```diff
@@ -7,6 +7,14 @@ TF_CONFIG_DIR=""
 TF_DEBUG="${TF_DEBUG:-0}"
 TF_ARGS=()
 ENVIRONMENT=""
+# shellcheck disable=2034
+VAULT_FORMAT="json"
+TF_USE_VAULT_AWS_STS=true
+AWS_STS_PATH="aws/sts/terraform_backend"
+TF_GENERATE_BACKEND=true
+TF_BACKEND_BUCKET_NAME="caascad-terraform-remote-state-bucket"
+TF_BACKEND_BUCKET_REGION_NAME="us-east-2"
```
Comment (yesteph) [inline tf:16]: Manage a TF_BACKEND_DYNAMODB_TABLE_NAME variable.

Comment (Benjile) [reply tf]: done

File: tf (hunk 5 above)
Comment (Benjile) [reply tf:269]: i added the dynamodb attribute

File: tf [hunk 7]
```diff
@@ -173,35 +187,130 @@ function _tf_init () {
...
+  INFRA_DOMAIN_NAME=$(_get_from_caascad_zones "${INFRA_ZONE_NAME}" "domain_name")
+  log_debug "infra domain name is: ${INFRA_ZONE_NAME}"
+  ACCOUNT_ID=$(_get_from_caascad_zones "${INFRA_ZONE_NAME}" "providers.aws.account_id")
+  log_debug "account_id is: ${ACCOUNT_ID}"
+  TF_BACKEND_BUCKET_NAME="${INFRA_ZONE_NAME}-${ACCOUNT_ID}-${TF_BACKEND_BUCKET_DEFAULT_SUFFIX}"
+  log_debug "backend bucket name is: ${TF_BACKEND_BUCKET_NAME}"
+  TF_BACKEND_DYNAMODB_TABLE_NAME="${INFRA_ZONE_NAME}-${ACCOUNT_ID}-tf-locks"
+  log_debug "dynamodb table name is: ${TF_BACKEND_DYNAMODB_TABLE_NAME}"
+
+  # set vault url, we must be able to override the vault url provided by caascad zones
```
Comment (eonpatapon) [inline tf:205]: Why ?

I think this will be a source of trouble if you use VAULT_ADDR of the user environment

File: tf [hunk 8]
```diff
@@ -173,35 +187,130 @@ function _tf_init () {
...
+}
+
+function _get_caascad_zones () {
+  log_debug "caascad zones url is: ${CAASCAD_ZONES_URL}"
+  if _is_local "${CAASCAD_ZONES_URL}"; then
+    log_debug "caascad zones files is local"
+    cp "${CAASCAD_ZONES_URL}" "${TF_CAASCAD_ZONES_FILE}"
+  else
+    log_debug "retrieve caascad zones file from ${CAASCAD_ZONES_URL}"
+    curl -s -o "${TF_CAASCAD_ZONES_FILE}" "${CAASCAD_ZONES_URL}"
```
Comment (eonpatapon) [inline tf:283]: Maybe add `--connect-timeout` and fail with an error message

File: tf [hunk 9]
```diff
@@ -173,35 +187,130 @@ function _tf_init () {
...
+  fi
+  log_debug "using caascad zones file in ${TF_CAASCAD_ZONES_FILE}"
+
+  jq -r -e --arg zone  "${ENVIRONMENT}" '.[$zone].name' < "${TF_CAASCAD_ZONES_FILE}" > /dev/null
+  # shellcheck disable=2181
+  if [[ $? -ne 0 ]]; then log_error "cannot find zone ${ENVIRONMENT} in caascad-zones.json file."; fi
```
Comment (eonpatapon) [inline tf:289]: This code won't be reached as we use `set -e`

Comment (eonpatapon) [inline tf:289]:
```suggestion
if ! jq -r -e --arg zone  "${ENVIRONMENT}" '.[$zone].name' < "${TF_CAASCAD_ZONES_FILE}" > /dev/null
then
  # shellcheck disable=2181
  log_error "cannot find zone ${ENVIRONMENT} in caascad-zones.json file."
fi
```

File: tf [hunk 10]
```diff
@@ -173,35 +187,130 @@ function _tf_init () {
...
+}
+
+function _get_from_caascad_zones() {
+  ZONE=$1
+  KEY=$2
+  VALUE=$(jq -r ".\"$ZONE\".$KEY" < "${TF_CAASCAD_ZONES_FILE}")
+  echo "$VALUE"
```
Comment (eonpatapon) [inline tf:296]:
```suggestion
  jq -r ".\"$ZONE\".$KEY" < "${TF_CAASCAD_ZONES_FILE}"
```

File: tf [hunk 11]
```diff
@@ -173,35 +187,130 @@ function _tf_init () {
...
     GIT_REVISION="local"
   else
     GIT_REVISION="${GIT_REVISION:-refs/heads/master}"
   fi
-
+  CAASCAD_ZONES_URL="${CAASCAD_ZONES_URL:-${CAASCAD_ZONES_URL_DEFAULT}}"
+  TF_USE_VAULT_AWS_STS="${TF_USE_VAULT_AWS_STS:-${TF_USE_VAULT_AWS_STS_DEFAULT}}"
+  TF_GENERATE_BACKEND="${TF_GENERATE_BACKEND:-${TF_GENERATE_BACKEND_DEFAULT}}"
+  TF_BACKEND_BUCKET_REGION_NAME="${TF_BACKEND_BUCKET_REGION_NAME:-${TF_BACKEND_BUCKET_DEFAULT_REGION_NAME}}"
+  TF_BACKEND_BUCKET_SUFFIX="${TF_BACKEND_BUCKET_SUFFIX:-${TF_BACKEND_BUCKET_DEFAULT_SUFFIX}}"
```
Comment (eonpatapon) [inline tf:313]: Why not putting all of this at the top then ?

File: tf [hunk 12]
```diff
@@ -173,35 +187,130 @@ function _tf_init () {
...
+  VALUE=$(jq -r ".\"$ZONE\".$KEY" < "${TF_CAASCAD_ZONES_FILE}")
+  echo "$VALUE"
 }
 
 function _tf_parsing () {
   # trying to source our environments variables
   # shellcheck disable=1091
   source "tffile" &>/dev/null || true
-  LIB_URL="${LIB_URL:-git@git.corp.cloudwatt.com:caascad/terraform/lib.git}"
+  LIB_URL="${LIB_URL:-${LIB_URL_DEFAULT}}"
```
Comment (eonpatapon) [inline tf:303]: Same as below

File: tf [hunk 13]
```diff
@@ -173,35 +187,130 @@ function _tf_init () {
...
+  _tf_generic init -upgrade=true "${init_args[@]}" >&2
+  BACKEND_CONFIG="$(_tf_backend_config)"
+  # we want word splitting here
+  # shellcheck disable=2086
+  _tf_generic init -upgrade=true ${BACKEND_CONFIG} >&2
+}
+
+function _tf_backend_config () {
+  if [[ ${TF_USE_VAULT_AWS_STS} -ne 0 ]]; then
+    token=$(VAULT_ADDR=${VAULT_ADDR} vault read "${AWS_STS_PATH}" -format=json)
```
Comment (eonpatapon) [inline tf:243]:
```suggestion
    token=$(vault read "${AWS_STS_PATH}" -format=json)
```

File: tf [hunk 14]
```diff
@@ -173,35 +187,130 @@ function _tf_init () {
...
+  TF_BACKEND_DYNAMODB_TABLE_NAME="${INFRA_ZONE_NAME}-${ACCOUNT_ID}-tf-locks"
+  log_debug "dynamodb table name is: ${TF_BACKEND_DYNAMODB_TABLE_NAME}"
+
+  # set vault url, we must be able to override the vault url provided by caascad zones
+  if [[ -z "${VAULT_ADDR}" ]]; then
+    VAULT_ADDR="https://vault.${INFRA_ZONE_NAME}.${INFRA_DOMAIN_NAME}"
+    log_debug "vault addr is: ${VAULT_ADDR}"
+  fi
+
+  # generate backend file
```
Comment (eonpatapon) [inline tf:211]: Not sure why it is optional ?

**Review by eonpatapon — CHANGES_REQUESTED** [review]

File: tf (hunk 7 above)
Comment (Benjile) [reply tf:205]: during a bootstrap process we should be able to override the VAULT_ADDR used

File: tf (hunk 13 above)
Comment (Benjile) [reply tf:263]: it won't work as we do not export this variable

Comment (Benjile) [reply tf:263]: i tested it :-)

File: tf (hunk 14 above)
Comment (Benjile) [reply tf:221]: a feature to not override a backend.tf provided in the configuration, or just to work locally

File: default.nix [hunk 15]
```diff
@@ -13,7 +13,7 @@ stdenv.mkDerivation rec {
   buildInputs = [ makeWrapper ];
   installPhase = ''
     install -m755 -D ${./tf} $out/bin/tf
-    wrapProgram $out/bin/tf --prefix PATH : "${lib.makeBinPath [ findutils gnused coreutils ]}"
+    wrapProgram $out/bin/tf --prefix PATH : "${lib.makeBinPath [ findutils gnused coreutils jq curl ]}"
```
Comment (eonpatapon) [inline default.nix:16]: vault missing

File: tf (hunk 7 above)
Comment (eonpatapon) [reply tf:214]: I woul prefer to allow to override it from tffile, but not from the env

**Review by eonpatapon — CHANGES_REQUESTED** [review]

File: tf (hunk 7 above)
Comment (Benjile) [reply tf:214]: at the beginning of the of the parse function we have a source tffile, so i will add before an unset VAULT_ADDR to fulfill your request

File: default.nix (hunk 15 above)
Comment (Benjile) [reply default.nix]: fixed

**Review by eonpatapon — APPROVED** [review]

**Pull request merged**